dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,369 @@
+                                                   size_t dtype_max_size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  uint8_t* shared_tensor_data = nullptr;
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].SetMeta(meta);
+      auto component_dtype_info = TypeTable::GetTypeInfo(dtype_[component_index]);
+      if (shared_tensor_data == nullptr) {
+        sample[component_index].reserve(align_up(size, dtype_max_size));
+        sample[component_index].Resize(size / component_dtype_info.size(), component_dtype_info);
+        shared_tensor_data = reinterpret_cast<uint8_t*>(sample[component_index].raw_mutable_data());
+        std::memset(shared_tensor_data + size, 0, align_up(size, dtype_max_size) - size);
+      } else {
+        sample[component_index].ShareData(
+            shared_tensor_data, size, {size / static_cast<int64_t>(component_dtype_info.size())},
+            component_dtype_info);
+      }
+      sample_was_set[component_index] = true;
+    }
+  }
+  return shared_tensor_data;
+}
+
+inline void WebdatasetLoader::MarkCached(std::vector<Tensor<CPUBackend>>& sample,
+                                         std::vector<bool>& sample_was_set,
+                                         const std::string& extension,
+                                         const std::string& source_info) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(true);
+
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].Reset();
+      sample[component_index].SetMeta(meta);
+      sample[component_index].set_type(TypeTable::GetTypeInfo(dtype_[component_index]));
+      sample[component_index].Resize({0});
+      sample_was_set[component_index] = true;
+    }
+  }
+}
+
+void WebdatasetLoader::ReadSample(vector<Tensor<CPUBackend>>& sample) {
+  MoveToNextShard(GetCurrentSampleIndex());
+  auto& current_wds_shard = wds_shards_[current_wds_shard_index_];
+  auto& current_sample = wds_shards_metadata_[current_wds_shard_index_][current_sample_index_];
+  current_wds_shard.SeekArchive(current_sample.start_offset);
+
+  vector<bool> sample_was_set(sample.size(), false);
+  while (current_wds_shard.TellArchive() < current_sample.end_offset) {
+    // Check in case of encountering a tar entry that is not a file
+    if (current_wds_shard.GetFileType() != detail::TarArchive::ENTRY_FILE) {
+      DALI_ENFORCE(current_wds_shard.NextFile(), "Index file reporting a file longer than actual");
+      continue;
+    }
+
+    // Check in case of encountering an unneeded entry
+    const std::string extension = GetExtension(current_wds_shard.GetFileName());
+    if (ext_map_.find(extension) == ext_map_.end()) {
+      DALI_ENFORCE(current_wds_shard.NextFile(), "Index file reporting a file longer than actual");
```

Comment:
Please adjust the error message. It would be good to print filename and offset.

Reply:
alright